File: jar-common/src/main/java/com/beancounter/common/json/CurrencyKeyDeserializer.java
package com.beancounter.common.json;

import com.beancounter.common.model.IsoCurrencyPair;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.KeyDeserializer;

public class CurrencyKeyDeserializer extends KeyDeserializer {
  @Override
  public IsoCurrencyPair deserializeKey(String key, DeserializationContext ctxt) {
    return IsoCurrencyPair.builder()
        .from(key.substring(0, 3))
        .to(key.substring(4, 7)).build();
  }
}

File: jar-common/src/main/java/com/beancounter/common/utils/CurrencyKeyDeserializer.kt
package com.beancounter.common.utils

import com.beancounter.common.model.IsoCurrencyPair
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.KeyDeserializer

class CurrencyKeyDeserializer : KeyDeserializer() {
    override fun deserializeKey(key: String, ctxt: DeserializationContext): IsoCurrencyPair {
        return IsoCurrencyPair(key.substring(0, 3), key.substring(4, 7))
    }
}
